### PR TITLE
adopt spotless for automated formatting

### PR DESCRIPTION
**Pull request by lemire**
fixes https://github.com/RoaringBitmap/RoaringBitmap/issues/757

### CODE REVIEW

**Comment by blacelle** [comment]
Having a quick look. I'm not very strong in Gradle, though I contribute a few bits into Spotless (which also has a mvn integration).

**Comment by blacelle** [comment]
I suspect some fishy interaction with the smelly 
```
tasks {
    register("build") {
        // dummy build task to appease release plugin
    }
}
```

We can expect some gentle & early feedback from Spotless people.

**Comment by lemire** [comment]
@blacelle I can remove this task, then it fails with...

```
☁  RoaringBitmapp [try_spotless] ./gradlew build

> Configure project :roaringbitmap
Project :roaringbitmap => 'roaringbitmap' Java module

FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':bsi:spotlessCheck'.
> Could not create task ':bsi:spotlessJavaCheck'.
   > Could not create task ':bsi:spotlessJava'.
      > Multiple steps with name 'google-java-format' for spotless format 'java'
```

**Comment by nedtwigg** [comment]
I think if you remove this chunk

```
tasks {
    register("build") {
        // dummy build task to appease release plugin
    }
}
```

then it will work. There are a few standard "built-in" tasks, I think `assemble`, `check`, `build`, and `clean`. They aren't technically built-in they are applied by the `base` plugin. Spotless (and many other plugins) access this convention by applying the base plugin

https://github.com/diffplug/spotless/blob/3638f8f9272fb14148436411bf26a4fb6503d13c/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessPlugin.java#L49-L50

but when that plugin runs it finds that you have already created the task in question. Just deleting that part of your build should work, or you can apply `base` yourself to create in a more ecosystem-compatible way.

**Comment by nedtwigg** [comment]
And then this

> Multiple steps with name 'google-java-format' for spotless format 'java'

is a known issue the `7.0.0` betas, it will be fixed in `7.0.0` final. Until then, you can either use `6.x` or comment-out the `removeUnusedImports`. Tracking issue for this problem:

https://github.com/diffplug/spotless/issues/2159

**Comment by lemire** [comment]
I followed @nedtwigg'd advice and ended up reformatting the whole code base.

**Comment by lemire** [comment]
Let me be more precise, our codebase was *already* properly formatted.

However, some sections were not. This will format it.

**Comment by blacelle** [comment]
Thanks a lot @nedtwigg !

**Comment by blacelle** [comment]
@lemire While we could progressively enrich Spotless configuration (e.g. to cover not-java files), I strongly suggest you add a `ratchetFrom` parameter: https://github.com/diffplug/spotless/tree/main/plugin-gradle#ratchet

The point is:

- without no ratchet, spotless checks the whole codebase all the time (potentially fine for an initial full reformatting)
- with ratchet, only the files knowing as having changed with the ratchet-ed branch (typically `master`) will be considered. A small con is it may lead to issues if your git stash holds removed files (e.g. files being removed, while considered as present by the `ratchet` feature).

**Comment by lemire** [comment]
Yeah, thanks @nedtwigg  :heart:

I plan to merge this once it goes green.

**Comment by lemire** [comment]
@blacelle 

Can you have a look at the diff that it introduces right now? It mostly just reformat the BSI files which is fine. This might cause a bit of pain for people hacking this part of the code, but it does not nearly change all files (which would be bad).

File: build.gradle.kts [hunk 1]
```diff
@@ -138,9 +139,14 @@ subprojects.filter { listOf("roaringbitmap", "bsi").contains(it.name) }.forEach
...
 
 
-tasks {
-    register("build") {
-        // dummy build task to appease release plugin
+subprojects {
+    apply(plugin = "com.diffplug.spotless")
+    spotless {
+        java {
+            googleJavaFormat("1.17.0").aosp().reflowLongStrings()
```
Comment (blacelle) [inline build.gradle.kts:146]: I doubt you want `aosp()`, especially as it leads to 4-spaces indentation (https://github.com/diffplug/spotless/issues/420#issuecomment-512579445)

**Comment by lemire** [comment]
@blacelle Ok. I am stopping this PR.

Would you just pick it up and tweak it?

**Comment by lemire** [comment]
@blacelle I have converted this PR to a draft. It works, but we won't merge it.

Can you contribute a better version?

**Comment by blacelle** [comment]
Let's do that. I 'll start with something slightly different like:

```

    spotless {
        ratchetFrom("origin/master")

        java {
            googleJavaFormat().reflowLongStrings()
            formatAnnotations()
        }
    }
```

There is still a bunch of changes fromcurrent code format. And `googleJavaFormat()` is not much tweakable.

I'll open a PR, just to push the wagon.

**Comment by lemire** [comment]
@blacelle This library is still used quite a bit, so making it sing is worth it.

**Comment by lemire** [comment]
Closing.

**Pull request closed without merging**